std: Stabilize the std::hash module

alexcrichton opened the pull request:
This commit aims to prepare the `std::hash` module for alpha by formalizing its
current interface whileholding off on adding `#[stable]` to the new APIs.  The
current usage with the `HashMap` and `HashSet` types is also reconciled by
separating out composable parts of the design. The primary goal of this slight
redesign is to separate the concepts of a hasher's state from a hashing
algorithm itself.

The primary change of this commit is to separate the `Hasher` trait into a
`Hasher` and a `HashState` trait. Conceptually the old `Hasher` trait was
actually just a factory for various states, but hashing had very little control
over how these states were used. Additionally the old `Hasher` trait was
actually fairly unrelated to hashing.

This commit redesigns the existing `Hasher` trait to match what the notion of a
`Hasher` normally implies with the following definition:

    trait Hasher {
        type Output;
        fn reset(&mut self);
        fn finish(&self) -> Output;
    }

This `Hasher` trait emphasizes that hashing algorithms may produce outputs other
than a `u64`, so the output type is made generic. Other than that, however, very
little is assumed about a particular hasher. It is left up to implementors to
provide specific methods or trait implementations to feed data into a hasher.

The corresponding `Hash` trait becomes:

    trait Hash<H: Hasher> {
        fn hash(&self, &mut H);
    }

The old default of `SipState` was removed from this trait as it's not something
that we're willing to stabilize until the end of time, but the type parameter is
always required to implement `Hasher`. Note that the type parameter `H` remains
on the trait to enable multidispatch for specialization of hashing for
particular hashers.

Note that `Writer` is not mentioned in either of `Hash` or `Hasher`, it is
simply used as part `derive` and the implementations for all primitive types.

With these definitions, the old `Hasher` trait is realized as a new `HashState`
trait in the `collections::hash_state` module as an unstable addition for
now. The current definition looks like:

    trait HashState {
        type Hasher: Hasher;
        fn hasher(&self) -> Hasher;
    }

The purpose of this trait is to emphasize that the one piece of functionality
for implementors is that new instances of `Hasher` can be created.  This
conceptually represents the two keys from which more instances of a
`SipHasher` can be created, and a `HashState` is what's stored in a
`HashMap`, not a `Hasher`.

Implementors of custom hash algorithms should implement the `Hasher` trait, and
only hash algorithms intended for use in hash maps need to implement or worry
about the `HashState` trait.

The entire module and `HashState` infrastructure remains `#[unstable]` due to it
being recently redesigned, but some other stability decision made for the
`std::hash` module are:

* The `Writer` trait remains `#[experimental]` as it's intended to be replaced
  with an `io::Writer` (more details soon).
* The top-level `hash` function is `#[unstable]` as it is intended to be generic
  over the hashing algorithm instead of hardwired to `SipHasher`
* The inner `sip` module is now private as its one export, `SipHasher` is
  reexported in the `hash` module.

And finally, a few changes were made to the default parameters on `HashMap`.

* The `RandomSipHasher` default type parameter was renamed to `RandomState`.
  This renaming emphasizes that it is not a hasher, but rather just state to
  generate hashers. It also moves away from the name "sip" as it may not always
  be implemented as `SipHasher`. This type lives in the
  `std::collections::hash_map` module as `#[unstable]`

* The associated `Hasher` type of `RandomState` is creatively called...
  `Hasher`! This concrete structure lives next to `RandomState` as an
  implemenation of the "default hashing algorithm" used for a `HashMap`. Under
  the hood this is currently implemented as `SipHasher`, but it draws an
  explicit interface for now and allows us to modify the implementation over
  time if necessary.

There are many breaking changes outlined above, and as a result this commit is
a:

[breaking-change]

rust-highfive commented:
r? @nikomatsakis

(rust_highfive has picked a reviewer for you, use r? to override)

alexcrichton commented:
This is a reopening of #19673, and does not currently yet build as it is waiting on a new snapshot. The first commit, however, was created in preparation for a new snapshot.

r? @aturon

Gankro commented:
CC @pczarn @thestinger

cgaebel commented:
HashMap changes LGTM

alexcrichton commented:
The commit message just gets longer! I've rebased with as many goodies as possible (those that have been implemented). This removes all the `#[old_impl_check]` tags related to hashing now that associated types are leveraged.

I've purged "sip" as a name from everything other than `SipHasher` itself, updated the commit message/PR description, and I think that this is ready to go.

r? @aturon

pczarn commented:
continuing from #19673

> When it comes to deriving, unless we hard-code one specific algorithm, I don't think we have any option other than incrementally updating a hash (e.g. just calling .hash(state) on all members).

@alexcrichton That's right, such an algorithm would have to be aware of generics. There is another approach to this problem: attempting to optimize these calls to .hash(state) while still doing them for all members.

> For example the &[u8] type likely cannot hash well due to the lack of specialization implementations, whereas the &str type can likely hash very well due to actually being one call to .write().

I think the impl of Hash for &str has two calls to .write().

> I'm sorry I didn't quite follow your comment, but are you basically saying that requiring .write() + .finish() is imposing overhead, even in the &str case for example? Are you also saying that this is not inherent to the SipHasher implementation, but rather inherent to the design of the hashing traits?

Perhaps using .finish() imposes inevitable overhead, but I wouldn't worry much about it. In my comment, I tried to make an implementation that would do much less work incrementally, yes, because @thestinger didn't support his ideas with a concrete design. Nothing came out of that. I have overestimated LLVM's understanding of pointer arithmetic optimization.

alexcrichton commented:
> I think the impl of Hash for &str has two calls to .write().

Ah yes good point, I forgot about the byte we include at the end!

> Perhaps using .finish() imposes inevitable overhead, but I wouldn't worry much about it.

Thanks for clarifying!

aturon commented on src/liballoc/rc.rs:
```diff
@@ -597,12 +597,20 @@ impl<T: Ord> Ord for Rc<T> {
     }
 }
+#[cfg(not(stage0))]
+impl<S: hash::Writer + hash::Hasher, T: Hash<S>> Hash<S> for Rc<T> {
```
Do you need the `hash::Writer` constraint here any longer?

aturon commented on src/liballoc/boxed.rs:
```diff
@@ -106,12 +106,20 @@ impl<T: ?Sized + Ord> Ord for Box<T> {
     }
 }
+#[cfg(not(stage0))]
+impl<S: hash::Writer + hash::Hasher, T: ?Sized + Hash<S>> Hash<S> for Box<T> {
```
Do you need the `hash::Writer` constraint here any longer?

aturon commented on src/libcollections/btree/map.rs:
```diff
@@ -827,6 +828,15 @@ impl<S: Writer, K: Hash<S>, V: Hash<S>> Hash<S> for BTreeMap<K, V> {
 }
+#[stable]
+#[cfg(not(stage0))]
+impl<S: Writer + Hasher, K: Hash<S>, V: Hash<S>> Hash<S> for BTreeMap<K, V> {
```
Do you need the `hash::Writer` constraint here any longer?

aturon commented on src/libcollections/dlist.rs:
```diff
@@ -675,7 +675,7 @@ impl<A: fmt::Show> fmt::Show for DList<A> {
 
 #[stable]
-impl<S: Writer, A: Hash<S>> Hash<S> for DList<A> {
+impl<S: Writer + Hasher, A: Hash<S>> Hash<S> for DList<A> {
```
`Writer` bound no longer needed.

alexcrichton replied:
I do think that this is needed because of the `self.len().hash(state)` b/c `uint` requires `Writer`

aturon commented on src/libcollections/ring_buf.rs:
```diff
@@ -1562,7 +1562,7 @@ impl<A: Ord> Ord for RingBuf<A> {
 
 #[stable]
-impl<S: Writer, A: Hash<S>> Hash<S> for RingBuf<A> {
+impl<S: Writer + Hasher, A: Hash<S>> Hash<S> for RingBuf<A> {
```
`Writer` bound no longer needed.

aturon commented on src/libcollections/string.rs:
```diff
@@ -820,12 +820,21 @@ impl fmt::Show for String {
 }
+#[experimental = "waiting on Hash stabilization"]
+#[cfg(not(stage0))]
+impl<H: hash::Writer + hash::Hasher> hash::Hash<H> for String {
```
`Writer` bound no *may* no longer be needed.

aturon commented on src/libcollections/vec.rs:
```diff
@@ -1183,12 +1183,20 @@ impl<T:Clone> Clone for Vec<T> {
     }
 }
+#[cfg(not(stage0))]
+impl<S: hash::Writer + hash::Hasher, T: Hash<S>> Hash<S> for Vec<T> {
```
`Writer` bound no longer needed.

aturon commented on src/libcollections/vec_map.rs:
```diff
@@ -85,7 +85,7 @@ impl<V:Clone> Clone for VecMap<V> {
 }
 
-impl<S: Writer, V: Hash<S>> Hash<S> for VecMap<V> {
+impl<S: Writer + Hasher, V: Hash<S>> Hash<S> for VecMap<V> {
```
`Writer` bound no longer needed?

aturon commented on src/libcore/hash/mod.rs:
```diff
@@ -46,8 +43,8 @@
 //!
-//! impl Hash for Person {
-//!     fn hash(&self, state: &mut SipState) {
+//! impl<H: Hasher + Writer> Hash<H> for Person {
```
`Writer` bound no longer needed.

aturon commented on src/libcore/hash/mod.rs:
```diff
@@ -56,186 +53,382 @@
+/// A hashable type.
+///
+/// The `H` type parameter is an abstract hash state that is used by the `Hash`
+/// to compute the hash.
```
Could add: `Specific implementations of `Hash` may impose their own requirements on `H` in order to be able to write into it.`

aturon commented on src/libcore/hash/mod.rs:
```diff
@@ -56,186 +53,382 @@
+///
+/// The specified value will be hashed with this hasher and then the resulting
+/// hash will be returned.
+pub fn hash<T: Hash<H>, H: Hasher + Default>(value: &T) -> H::Output {
```
Hmm, have you tried this out? It seems like you'll have to explicitly specify the `H` argument in most cases?

alexcrichton replied:
Yeah, although there may perhaps be some better methods of providing a function like this:

* We could provide a default method on the `Hash` trait with a `where Hasher: Default` clause is specified and put this function there. That way you'd say something like `foo.hash_value::<Algorithm>()`
* I *think* that with Niko's default type parameter RFC we could swap the order of `T` and `H`, and you would only have to specify `H`, right? That is, `foo::<H>` is equivalent to `foo::<H, _>`?

This function isn't super useful now that the default `SipHasher` bound was removed, so it may be better to just remove it.

aturon commented on src/libcore/hash/mod.rs:
```diff
@@ -56,186 +53,382 @@
-                state.write(&[]);
+    macro_rules! impl_hash_tuple {
+        () => (
+            impl<S: Writer> Hash<S> for () {
```
This could probably be rewritten to not require `Writer`.

aturon commented on src/libcore/hash/mod.rs:
```diff
@@ -56,186 +53,382 @@
+        );
+
+        ( $($name:ident)+) => (
+            impl<S: Writer, $($name: Hash<S>),*> Hash<S> for ($($name,)*) {
```
Can drop `Writer`.

aturon commented on src/libcore/hash/mod.rs:
```diff
@@ -56,186 +53,382 @@
+    impl_hash_tuple! { A B C D E F G H I J K }
+    impl_hash_tuple! { A B C D E F G H I J K L }
+
+    impl<S: Writer, T: Hash<S>> Hash<S> for [T] {
```
Can drop `Writer`.

aturon commented on src/libcore/hash/mod.rs:
```diff
@@ -56,186 +53,382 @@
-        self.len().hash(state);
-        for elt in self.iter() {
-            elt.hash(state);
+    impl<'a, S: Writer, T: ?Sized + Hash<S>> Hash<S> for &'a T {
```
Can drop `Writer`.

aturon commented on src/libcore/hash/mod.rs:
```diff
@@ -56,186 +53,382 @@
     }
-}
 
+    impl<'a, S: Writer, T: ?Sized + Hash<S>> Hash<S> for &'a mut T {
```
Can drop `Writer`.

aturon commented on src/libcore/hash/mod.rs:
```diff
@@ -56,186 +53,382 @@
-    #[inline]
-    fn hash(&self, state: &mut S) {
-        (**self).hash(state);
+    impl<S: Writer, T> Hash<S> for *const T {
```
Can drop `Writer`.

aturon replied:
Ooops, nevermind.

aturon commented on src/libcore/hash/mod.rs:
```diff
@@ -56,186 +53,382 @@
-        Hash::hash(&**self, state)
+    macro_rules! impl_hash_tuple {
+        () => (
+            impl<S: Writer + Hasher> Hash<S> for () {
```
Should be able to rewrite to avoid `Writer` bound.

aturon commented on src/libcore/hash/mod.rs:
```diff
@@ -56,186 +53,382 @@
+            impl<S: Writer + Hasher, $($name: Hash<S>),*> Hash<S> for ($($name,)*) {
+                #[inline]
+                #[allow(non_snake_case)]
+                fn hash(&self, state: &mut S) {
```
Can drop `Writer`.

aturon commented on src/libcore/hash/mod.rs:
```diff
@@ -56,186 +53,382 @@
+    impl_hash_tuple! { A B C D E F G H I J K L }
+
+    impl<S: Writer + Hasher, T: Hash<S>> Hash<S> for [T] {
+        #[inline]
```
Can drop `Writer`.

aturon commented on src/libcore/hash/mod.rs:
```diff
@@ -56,186 +53,382 @@
+    }
+
+
+    impl<'a, S: Writer + Hasher, T: ?Sized + Hash<S>> Hash<S> for &'a T {
```
Can drop `Writer`.

aturon commented on src/libcore/hash/mod.rs:
```diff
@@ -56,186 +53,382 @@
+        }
+    }
+
+    impl<'a, S: Writer + Hasher, T: ?Sized + Hash<S>> Hash<S> for &'a mut T {
```
Can drop `Writer`.

aturon commented on src/libcore/hash/mod.rs:
```diff
@@ -56,186 +53,382 @@
+        }
+    }
+
+    impl<'a, T, B: ?Sized, S: Hasher> Hash<S> for Cow<'a, T, B>
```
Can drop `Writer`.

aturon commented on src/libcore/hash/sip.rs:
```diff
@@ -195,355 +164,60 @@ impl Writer for SipState {
-}
-
-#[cfg(test)]
-mod tests {
```
Looks like we're dropping this entire thing? Why?

alexcrichton replied:
These tests are currently duplicated with `src/libcoretest/hash/sip.rs`, or I may have moved them there, I kinda forget which! Either way they do still exist :)

aturon commented on src/librustc/lint/mod.rs:
```diff
@@ -186,7 +186,7 @@ impl PartialEq for LintId {
 impl Eq for LintId { }
 
-impl<S: hash::Writer> hash::Hash<S> for LintId {
+impl<S: hash::Writer + hash::Hasher> hash::Hash<S> for LintId {
```
Can drop `Writer`.

aturon commented on src/librustc/metadata/decoder.rs:
```diff
@@ -94,7 +93,7 @@ pub fn maybe_find_item<'a>(item_id: ast::NodeId,
     lookup_hash(items,
                 |a| eq_item(a, item_id),
-                hash::hash(&(item_id as i64)))
+                hash::hash::<i64, SipHasher>(&(item_id as i64)))
```
Ahha, so yes you need the args.

aturon commented on src/librustc/middle/ty.rs:
```diff
@@ -946,11 +946,18 @@ impl<'tcx> PartialEq for TyS<'tcx> {
     }
 }
+#[cfg(not(stage0))]
+impl<'tcx, S: Writer + Hasher> Hash<S> for TyS<'tcx> {
```
Can drop `Writer`.

aturon commented on src/librustc/middle/ty.rs:
```diff
@@ -968,7 +975,7 @@ impl<'tcx> PartialEq for InternedTy<'tcx> {
 impl<'tcx> Eq for InternedTy<'tcx> {}
 
-impl<'tcx, S: Writer> Hash<S> for InternedTy<'tcx> {
+impl<'tcx, S: Writer + Hasher> Hash<S> for InternedTy<'tcx> {
```
May be able to drop `Writer`

aturon commented on src/libstd/collections/hash/map.rs:
```diff
@@ -1031,7 +1038,7 @@ impl<K: Eq + Hash<S>, V, S, H: Hasher<S>> HashMap<K, V, H> {
     #[stable]
     pub fn get<Q: ?Sized>(&self, k: &Q) -> Option<&V>
-        where Q: Hash<S> + Eq + BorrowFrom<K>
+        where Q: Hash<<S as HashState>::Hasher> + Eq + BorrowFrom<K>
```
;_;

aturon commented on src/libstd/collections/hash/map.rs:
```diff
@@ -512,14 +516,16 @@ impl<K: Hash + Eq, V> HashMap<K, V, RandomSipHasher> {
 
-#[old_impl_check]
-impl<K: Eq + Hash<S>, V, S, H: Hasher<S>> HashMap<K, V, H> {
+impl<K, V, S> HashMap<K, V, S>
```
I suggest adding an `H` parameter and making `S: HashState<Hasher = H>`

aturon commented on src/libstd/collections/hash/map.rs:
```diff
@@ -1189,10 +1196,12 @@ fn search_entry_hashed<'a, K: Eq, V>(table: &'a mut RawTable<K,V>, hash: SafeHas
-#[old_impl_check]
-impl<K: Eq + Hash<S>, V: PartialEq, S, H: Hasher<S>> PartialEq for HashMap<K, V, H> {
-    fn eq(&self, other: &HashMap<K, V, H>) -> bool {
+impl<K, V, S> PartialEq for HashMap<K, V, S>
```
Same suggestion re: factoring out an `H` with `S: HashState<Hasher=H>`

aturon commented on src/libstd/collections/hash/map.rs:
```diff
@@ -1202,12 +1211,18 @@ impl<K: Eq + Hash<S>, V: PartialEq, S, H: Hasher<S>> PartialEq for HashMap<K, V,
 #[stable]
-#[old_impl_check]
-impl<K: Eq + Hash<S>, V: Eq, S, H: Hasher<S>> Eq for HashMap<K, V, H> {}
+impl<K, V, S> Eq for HashMap<K, V, S>
```
Same suggestion re: factoring out an `H` with `S: HashState<Hasher=H>`

aturon commented on src/libstd/collections/hash/map.rs:
```diff
@@ -1202,12 +1211,18 @@ impl<K: Eq + Hash<S>, V: PartialEq, S, H: Hasher<S>> PartialEq for HashMap<K, V,
 #[stable]
-#[old_impl_check]
-impl<K: Eq + Hash<S> + Show, V: Show, S, H: Hasher<S>> Show for HashMap<K, V, H> {
+impl<K, V, S> Show for HashMap<K, V, S>
```
Same suggestion re: factoring out an `H` with `S: HashState<Hasher=H>`

aturon commented on src/libstd/collections/hash/map.rs:
```diff
@@ -1221,18 +1236,22 @@ impl<K: Eq + Hash<S> + Show, V: Show, S, H: Hasher<S>> Show for HashMap<K, V, H>
-    #[stable]
-    fn default() -> HashMap<K, V, H> {
-        HashMap::with_hasher(Default::default())
+impl<K, V, S> Default for HashMap<K, V, S>
```
Same suggestion re: factoring out an `H` with `S: HashState<Hasher=H>`

aturon commented on src/libstd/collections/hash/map.rs:
```diff
@@ -1221,18 +1236,22 @@ impl<K: Eq + Hash<S> + Show, V: Show, S, H: Hasher<S>> Show for HashMap<K, V, H>
-#[old_impl_check]
-impl<K: Hash<S> + Eq, Q: ?Sized, V, S, H: Hasher<S>> Index<Q> for HashMap<K, V, H>
-    where Q: BorrowFrom<K> + Hash<S> + Eq
+impl<K, Q: ?Sized, V, S> Index<Q> for HashMap<K, V, S>
```
Same suggestion re: factoring out an `H` with `S: HashState<Hasher=H>`

aturon commented on src/libstd/collections/hash/map.rs:
```diff
@@ -1243,9 +1262,11 @@ impl<K: Hash<S> + Eq, Q: ?Sized, V, S, H: Hasher<S>> Index<Q> for HashMap<K, V,
-impl<K: Hash<S> + Eq, Q: ?Sized, V, S, H: Hasher<S>> IndexMut<Q> for HashMap<K, V, H>
-    where Q: BorrowFrom<K> + Hash<S> + Eq
+impl<K, V, S, Q: ?Sized> IndexMut<Q> for HashMap<K, V, S>
+    where K: Eq + Hash<<S as HashState>::Hasher>,
```
Same suggestion re: factoring out an `H` with `S: HashState<Hasher=H>`

aturon commented on src/libstd/collections/hash/map.rs:
```diff
@@ -1473,19 +1494,26 @@ impl<'a, K: 'a, V: 'a> VacantEntry<'a, K, V> {
-impl<K: Eq + Hash<S>, V, S, H: Hasher<S> + Default> FromIterator<(K, V)> for HashMap<K, V, H> {
-    fn from_iter<T: Iterator<Item=(K, V)>>(iter: T) -> HashMap<K, V, H> {
+impl<K, V, S> FromIterator<(K, V)> for HashMap<K, V, S>
+    where K: Eq + Hash<<S as HashState>::Hasher>,
```
Same suggestion re: factoring out an `H` with `S: HashState<Hasher=H>`

aturon commented on src/libstd/collections/hash/map.rs:
```diff
@@ -1473,19 +1494,26 @@ impl<'a, K: 'a, V: 'a> VacantEntry<'a, K, V> {
-#[old_impl_check]
-impl<K: Eq + Hash<S>, V, S, H: Hasher<S>> Extend<(K, V)> for HashMap<K, V, H> {
+impl<K, V, S> Extend<(K, V)> for HashMap<K, V, S>
+    where K: Eq + Hash<<S as HashState>::Hasher>,
```
Same suggestion re: factoring out an `H` with `S: HashState<Hasher=H>`

aturon commented on src/libstd/collections/hash/set.rs:
```diff
@@ -120,13 +121,16 @@ impl<T: Hash + Eq> HashSet<T, RandomSipHasher> {
-#[old_impl_check]
-impl<T: Eq + Hash<S>, S, H: Hasher<S>> HashSet<T, H> {
+impl<T, S> HashSet<T, S>
+    where T: Eq + Hash<<S as HashState>::Hasher>,
```
Same suggestion re: factoring out an `H` with `S: HashState<Hasher=H>`

aturon commented on src/libstd/collections/hash/set.rs:
```diff
@@ -579,12 +589,18 @@ impl<T: Eq + Hash<S>, S, H: Hasher<S>> PartialEq for HashSet<T, H> {
-#[old_impl_check]
-impl<T: Eq + Hash<S>, S, H: Hasher<S>> Eq for HashSet<T, H> {}
+impl<T, S> Eq for HashSet<T, S>
+    where T: Eq + Hash<<S as HashState>::Hasher>,
```
Same suggestion re: factoring out an `H` with `S: HashState<Hasher=H>`

aturon commented on src/libstd/collections/hash/set.rs:
```diff
@@ -579,12 +589,18 @@ impl<T: Eq + Hash<S>, S, H: Hasher<S>> PartialEq for HashSet<T, H> {
-#[old_impl_check]
-impl<T: Eq + Hash<S> + fmt::Show, S, H: Hasher<S>> fmt::Show for HashSet<T, H> {
+impl<T, S> fmt::Show for HashSet<T, S>
+    where T: Eq + Hash<<S as HashState>::Hasher> + fmt::Show,
```
Same suggestion re: factoring out an `H` with `S: HashState<Hasher=H>`

aturon commented on src/libstd/collections/hash/set.rs:
```diff
@@ -598,19 +614,25 @@ impl<T: Eq + Hash<S> + fmt::Show, S, H: Hasher<S>> fmt::Show for HashSet<T, H> {
-impl<T: Eq + Hash<S>, S, H: Hasher<S> + Default> FromIterator<T> for HashSet<T, H> {
-    fn from_iter<I: Iterator<Item=T>>(iter: I) -> HashSet<T, H> {
+impl<T, S> FromIterator<T> for HashSet<T, S>
+    where T: Eq + Hash<<S as HashState>::Hasher>,
```
Same suggestion re: factoring out an `H` with `S: HashState<Hasher=H>`

aturon commented on src/libstd/collections/hash/set.rs:
```diff
@@ -598,19 +614,25 @@ impl<T: Eq + Hash<S> + fmt::Show, S, H: Hasher<S>> fmt::Show for HashSet<T, H> {
-#[old_impl_check]
-impl<T: Eq + Hash<S>, S, H: Hasher<S>> Extend<T> for HashSet<T, H> {
+impl<T, S> Extend<T> for HashSet<T, S>
+    where T: Eq + Hash<<S as HashState>::Hasher>,
```
Same suggestion re: factoring out an `H` with `S: HashState<Hasher=H>`

aturon commented on src/libstd/collections/hash/set.rs:
```diff
@@ -619,21 +641,26 @@ impl<T: Eq + Hash<S>, S, H: Hasher<S>> Extend<T> for HashSet<T, H> {
-#[old_impl_check]
-impl<T: Eq + Hash<S>, S, H: Hasher<S> + Default> Default for HashSet<T, H> {
+impl<T, S> Default for HashSet<T, S>
+    where T: Eq + Hash<<S as HashState>::Hasher>,
```
Same suggestion re: factoring out an `H` with `S: HashState<Hasher=H>`

aturon commented on src/libstd/collections/hash/set.rs:
```diff
@@ -619,21 +641,26 @@ impl<T: Eq + Hash<S>, S, H: Hasher<S>> Extend<T> for HashSet<T, H> {
-BitOr<&'b HashSet<T, H>> for &'a HashSet<T, H> {
-    type Output = HashSet<T, H>;
+impl<'a, 'b, T, S> BitOr<&'b HashSet<T, S>> for &'a HashSet<T, S>
+    where T: Eq + Hash<<S as HashState>::Hasher> + Clone,
```
Same suggestion re: factoring out an `H` with `S: HashState<Hasher=H>`

aturon commented on src/libstd/collections/hash/set.rs:
```diff
@@ -653,18 +680,20 @@ BitOr<&'b HashSet<T, H>> for &'a HashSet<T, H> {
-BitAnd<&'b HashSet<T, H>> for &'a HashSet<T, H> {
-    type Output = HashSet<T, H>;
+impl<'a, 'b, T, S> BitAnd<&'b HashSet<T, S>> for &'a HashSet<T, S>
+    where T: Eq + Hash<<S as HashState>::Hasher> + Clone,
```
Same suggestion re: factoring out an `H` with `S: HashState<Hasher=H>`

aturon commented on src/libstd/collections/hash/set.rs:
```diff
@@ -684,18 +713,20 @@ BitAnd<&'b HashSet<T, H>> for &'a HashSet<T, H> {
-BitXor<&'b HashSet<T, H>> for &'a HashSet<T, H> {
-    type Output = HashSet<T, H>;
+impl<'a, 'b, T, S> BitXor<&'b HashSet<T, S>> for &'a HashSet<T, S>
+    where T: Eq + Hash<<S as HashState>::Hasher> + Clone,
```
Same suggestion re: factoring out an `H` with `S: HashState<Hasher=H>`

aturon commented on src/libstd/collections/hash/set.rs:
```diff
@@ -715,18 +746,20 @@ BitXor<&'b HashSet<T, H>> for &'a HashSet<T, H> {
-Sub<&'b HashSet<T, H>> for &'a HashSet<T, H> {
-    type Output = HashSet<T, H>;
+impl<'a, 'b, T, S> Sub<&'b HashSet<T, S>> for &'a HashSet<T, S>
+    where T: Eq + Hash<<S as HashState>::Hasher> + Clone,
```
Same suggestion re: factoring out an `H` with `S: HashState<Hasher=H>`

aturon commented on src/libstd/collections/hash/set.rs:
```diff
@@ -824,9 +857,10 @@ impl<'a, K: 'a> Iterator for Drain<'a, K> {
-impl<'a, T, S, H> Iterator for Intersection<'a, T, H>
-    where T: Eq + Hash<S>, H: Hasher<S>
+impl<'a, T, S> Iterator for Intersection<'a, T, S>
+    where T: Eq + Hash<<S as HashState>::Hasher>,
```
Same suggestion re: factoring out an `H` with `S: HashState<Hasher=H>`

aturon commented on src/libstd/collections/hash/set.rs:
```diff
@@ -848,9 +882,10 @@ impl<'a, T, S, H> Iterator for Intersection<'a, T, H>
-impl<'a, T, S, H> Iterator for Difference<'a, T, H>
-    where T: Eq + Hash<S>, H: Hasher<S>
+impl<'a, T, S> Iterator for Difference<'a, T, S>
+    where T: Eq + Hash<<S as HashState>::Hasher>,
```
Same suggestion re: factoring out an `H` with `S: HashState<Hasher=H>`

aturon commented on src/libstd/collections/hash/set.rs:
```diff
@@ -872,9 +907,10 @@ impl<'a, T, S, H> Iterator for Difference<'a, T, H>
-impl<'a, T, S, H> Iterator for SymmetricDifference<'a, T, H>
-    where T: Eq + Hash<S>, H: Hasher<S>
+impl<'a, T, S> Iterator for SymmetricDifference<'a, T, S>
+    where T: Eq + Hash<<S as HashState>::Hasher>,
```
Same suggestion re: factoring out an `H` with `S: HashState<Hasher=H>`

aturon commented on src/libstd/collections/hash/set.rs:
```diff
@@ -883,9 +919,10 @@ impl<'a, T, S, H> Iterator for SymmetricDifference<'a, T, H>
-impl<'a, T, S, H> Iterator for Union<'a, T, H>
-    where T: Eq + Hash<S>, H: Hasher<S>
+impl<'a, T, S> Iterator for Union<'a, T, S>
+    where T: Eq + Hash<<S as HashState>::Hasher>,
```
Same suggestion re: factoring out an `H` with `S: HashState<Hasher=H>`

aturon commented on src/libstd/collections/hash/table.rs:
```diff
@@ -138,12 +139,18 @@ impl SafeHash {
 /// module to generate a SafeHash.
-pub fn make_hash<T: ?Sized + Hash<S>, S, H: Hasher<S>>(hasher: &H, t: &T) -> SafeHash {
+pub fn make_hash<T: ?Sized, S>(hash_state: &S, t: &T) -> SafeHash
+    where T: Hash< <S as HashState>::Hasher>,
```
Same suggestion re: factoring out an `H` with `S: HashState<Hasher=H>`

alexcrichton commented:
Ok, I believe I took care of all unnecessary `Writer` bounds as well as the formatting in `map.rs`/`set.rs`. I left a bunch of `Writer` bounds because the implementations require it one way or another (they might hash a primitive, for example), but implementations like pointers just forward the `H` parameter though.

aturon commented:
Awesome work, @alexcrichton
